common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,224 @@
+        "AggregateRetentionFunction"
+    }
+
+    fn return_type(&self) -> Result<DataTypePtr> {
+        Ok(JsonValue::to_data_type())
+    }
+
+    fn init_state(&self, place: StateAddr) {
+        place.write(|| AggregateRetentionState { events: 0 });
+    }
+
+    fn state_layout(&self) -> std::alloc::Layout {
+        Layout::new::<AggregateRetentionState>()
+    }
+
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        _validity: Option<&common_arrow::arrow::bitmap::Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for i in 0..input_rows {
+            for j in 0..self.events_size {
+                if new_columns[j as usize].get_data(i) {
+                    state.add(j);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn accumulate_row(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        row: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for j in 0..self.events_size {
+            if new_columns[j as usize].get_data(row) {
+                state.add(j);
+            }
+        }
+        Ok(())
+    }
+
+    fn serialize(&self, place: StateAddr, writer: &mut BytesMut) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.serialize(writer)
+    }
+
+    fn deserialize(&self, place: StateAddr, reader: &mut &[u8]) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.deserialize(reader)
+    }
+
+    fn merge(&self, place: StateAddr, rhs: StateAddr) -> Result<()> {
+        let rhs = rhs.get::<AggregateRetentionState>();
+        let state = place.get::<AggregateRetentionState>();
+        state.merge(rhs);
+        Ok(())
+    }
+
+    #[allow(unused_mut)]
+    fn merge_result(
+        &self,
+        place: StateAddr,
+        array: &mut dyn common_datavalues::MutableColumn,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let builder: &mut MutableObjectColumn<JsonValue> = Series::check_get_mutable_column(array)?;
+        let mut vec: Vec<JsonValue> = vec![json!(0u8); self.events_size as usize];
+        if state.events & 1 == 1 {
+            vec[0] = json!(1u8);
+            for i in 1..self.events_size {
+                if state.events & (1 << i) != 0 {
+                    vec[i as usize] = json!(1u8);
+                }
+            }
+        }
+        builder.append_value(JsonValue::Array(vec));
```

Comment:
we can use a `Vec<u8>` vector and convert to `JsonValue` here, instead of converting each element to `JsonValue`

Reply:
good advice!